Projects/Shapes/Shapes/Shapes/DynamicArray.h

```diff
@@ -0,0 +1,341 @@
+	bool ShrinkToFit()
+	{
+		if (Capacity == Size)
+		{
+			return false;
+		}
+
+		T* NewArray = new T[Size];
+		for (int i = 0; i < Size; i++)
+		{
+			NewArray[i] = Array[i];
+		}
+
+		delete[] Array;
+		Array = NewArray;
+
+		Capacity = Size;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//											MODIFIERS
+	//-------------------------------------------------------------------------------------------
+
+	void Clear()
+	{
+		Size = 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PushBack(const T& Item)
+	{
+		if (Size == Capacity)
+		{
+			if (Capacity == 0)
+			{
+				Reserve(1);
+			}
+			Reserve(Capacity * 2);
+		}
+		Array[Size] = Item;
+		Size++;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PopBack()
+	{
+		if (Size > 0)
+		{
+			Size--;
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Insert(const T& Item, const int Index)
+	{
+		if (Index > Size)
+		{
+			return false;
+		}
+
+		if (Index == Size)
+		{
+			PushBack(Item);
+			Size++;
+			return true;
+		}
+
+		if (Size == Capacity)
+		{
+			Reserve(Capacity * 2);
+		}
+		
+		for (int i = Size; i > Index; i--)
+		{
+			Array[i] = Array[i-1];
+		}
+
+		Array[Index] = Item;
+		Size++;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Erase(const int Index)
+	{
+		if (Index > Size)
+		{
+			return false;
+		}
+
+		if (Index == Size)
+		{
+			PopBack();
+			return true;
+		}
+
+		for (int i = 0; i < Size - 1; i++)
+		{
+			Array[i] = Array[i + 1];
+		}
+		Size--;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Resize(const int NewSize)
+	{
+		if (Size > Capacity)
+		{
+			Reserve(NewSize);
+		}
+		Size = NewSize;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Resize(const int NewSize, const T& Item)
+	{
+		if (Size <= NewSize)
+		{
+			Size = NewSize;
+		}
+		else
+		{
+			if (NewSize > Capacity)
+			{
+				Reserve(NewSize);
+			}
+			for (int i = Size; i < NewSize; i++)
+			{
+				Array[i] = Item;
+			}
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Append(const TDynamicArray<T>& OtherArray)
+	{
+		Reserve(Size + OtherArray.GetSize());
+
+		for (int i = 0; i < OtherArray.GetSize(); i++)
+		{
+			PushBack(OtherArray[i]);
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//											CUSTOM METHODS
+	//-------------------------------------------------------------------------------------------
+
+	void RefactorCapacityTo(const int NewCapacity)
```

Comment:
Hmmm looks good but unnecessarily complicated. A resize and a shrink-to-fit would do the trick.

Reply:
I can see how it would work, would it be something similar to this?

Say you have a vector of items called Inventory, which had its initial capacity as 30 since that's the maximum capacity at the start of the game. After an unfortunate accident, the player now has their inventory reduced to a maximum of 10. 

```
//Game Start
TDynamicArray<Item> Inventory(30);
...
//Unfortunate Accident
int PreviousSize = Inventory.GetSize();
Inventory.Resize(10);
Inventory.ShrinkToFit(10);
Inventory.Resize(PreviousSize);
```

I assume that you would have to store the previous size, as you would want the size to be correct after the capacity change. Or is there a better way? The "better way" that I thought was this method, but I could be wrong: I remember when browsing online someone mentioned doing this with a swap, but I wasn't too sure how that changed things.